src/passes/Print.cpp

```diff
@@ -119,59 +131,104 @@ std::ostream& operator<<(std::ostream& os, SigName sigName) {
             }
             os << "]";
             continue;
+          } else if (heapType.isStruct()) {
+            auto struct_ = heapType.getStruct();
+            os << "{";
+            auto sep = "";
+            for (auto& field : struct_.fields) {
+              os << sep;
+              sep = "_";
+              if (field.mutable_) {
+                os << "mut:";
+              }
+              printType(field.type);
+            }
+            os << "}";
+            continue;
+          } else if (heapType.isArray()) {
+            os << "[";
+            auto element = heapType.getArray().element;
+            if (element.mutable_) {
+              os << "mut:";
+            }
+            printType(element.type);
+            os << "]";
+            continue;
           }
         }
         os << t;
       }
     }
   };
 
+  auto type = typeName.type;
+  if (type.isBasic()) {
+    os << type;
+    return os;
+  }
   os << '$';
-  printType(sigName.sig.params);
-  os << "_=>_";
-  printType(sigName.sig.results);
+  if (type.isSignature()) {
+    auto sig = type.getSignature();
+    printType(sig.params);
+    os << "_=>_";
+    printType(sig.results);
+  } else if (type.isStruct()) {
+    auto struct_ = type.getStruct();
+    os << "{";
+    auto sep = "";
+    for (auto& field : struct_.fields) {
+      os << sep;
+      sep = "_";
+      if (field.mutable_) {
+        os << "mut:";
+      }
+      printType(field.type);
+    }
+    os << "}";
+  } else if (type.isArray()) {
+    os << "[";
+    auto element = type.getArray().element;
+    if (element.mutable_) {
+      os << "mut:";
+    }
+    printType(element.type);
+    os << "]";
+  } else {
+    os << type;
+  }
```

Comment:
I think contents of this function overlaps with the `printType` lambda function above... Can we unify them?

Reply:
I think this is the same as @tlively 's comment above? Yes, there is duplication with some changes. I'd like to refactor them together in a followup, if that's ok.

Reply:
(should be fixed now)